File: C#/src/Implementations/HpManager.cs
public class HpManager : IHpManager
{
    private GameConfig gameConfig;
    private readonly int MAXHP;
    private int currentHP;

    public HpManager(GameConfig gameConfig)
    {
        this.gameConfig = gameConfig;
        MAXHP = gameConfig.maxPlayerHP;
        currentHP = gameConfig.maxPlayerHP;
    }

    public void LoseHP(GameState gameState)
    {
        int damage = gameState.gameStateManager.GetDamage();
        if (currentHP > 0)
        {
            SetHP(GetHP() - damage);
        }
    }

    public void RegainHP()
    {
        if (currentHP > 0 && currentHP < MAXHP)
            currentHP++;
    }

    public void ResetHP()
    {
        currentHP = MAXHP;
    }

    public int GetHP()
    {
        return currentHP;
    }

    public void SetHP(int newHP)
    {
        currentHP = newHP;
    }
}

File: C++/src/Implementations/HpManager.cpp
#include "HpManager.h"
#include <cstdint>

void HpManager::SetHP(uint8_t newHP) { currentHP = newHP; }

void HpManager::LoseHP(uint8_t damage) { if (currentHP > 0) { currentHP -= damage; } }

void HpManager::RegainHP() { if (currentHP > 0 && currentHP < maxHP) { currentHP++; } }

void HpManager::ResetHP() { currentHP = GameConfig::maxPlayerHP; }

int8_t HpManager::GetHP() const { return currentHP; }